### ISSUE TITLE
Training loss keeps increasing when I removed the distribute computation code and trained with Single GPU

### ISSUE THREAD

**Issue by ftongss** [zhu-xlab/SSL4EO-S12#8]
Hello,

   Thanks for your talent work.
   
   Currently I want to recreate the training of the network with Sentinel 2-L2C images.

   As I only have one GPU, I removed the code about distribute computation in [https://github.com/zhu-xlab/SSL4EO-S12/blob/main/src/benchmark/pretrain_ssl/pretrain_moco_v2_s2c.py](url). Also , I revised the code for shuffle batch normalization (line 156~162 in [https://github.com/zhu-xlab/SSL4EO-S12/blob/main/src/benchmark/pretrain_ssl/models/moco/builder.py](url) ) .

  After revising the code, I can run the training of the model. However, **I can't see any improvement of the training loss**. The training loss just kept increasing when the training epochs increased. 

  At first, I trained the model with the 100 patches subset (`epochs= 100, batch_size = 32` ) and got the following training process :

  

> Use GPU: 0 for training
> 
> => creating model 'resnet50'
> model created.
> start training...
> Epoch: [0][0/3]    Time  5.731 ( 5.731)    Data  0.957 ( 0.957)    Loss 6.5629e+00 (6.5629e+00)    Acc@1 100.00 (100.00)    Acc@5 100.00 (100.00)
> 0
> Epoch: [1][0/3]    Time  0.872 ( 0.872)    Data  0.705 ( 0.705)    Loss 6.7108e+00 (6.7108e+00)    Acc@1   0.00 (  0.00)    Acc@5   3.12 (  3.12)
> 1
> Epoch: [2][0/3]    Time  0.833 ( 0.833)    Data  0.665 ( 0.665)    Loss 6.7479e+00 (6.7479e+00)    Acc@1   0.00 (  0.00)    Acc@5   3.12 (  3.12)
> 2
> Epoch: [3][0/3]    Time  0.798 ( 0.798)    Data  0.630 ( 0.630)    Loss 6.8232e+00 (6.8232e+00)    Acc@1   0.00 (  0.00)    Acc@5   3.12 (  3.12)
> 3
> Epoch: [4][0/3]    Time  0.859 ( 0.859)    Data  0.691 ( 0.691)    Loss 6.8701e+00 (6.8701e+00)    Acc@1   0.00 (  0.00)    Acc@5   3.12 (  3.12)
> 4
> Epoch: [5][0/3]    Time  0.820 ( 0.820)    Data  0.651 ( 0.651)    Loss 6.9370e+00 (6.9370e+00)    Acc@1   3.12 (  3.12)    Acc@5   9.38 (  9.38)
> 5
> ...
> Epoch: [95][0/3]    Time  0.734 ( 0.734)    Data  0.563 ( 0.563)    Loss 8.8116e+00 (8.8116e+00)    Acc@1   9.38 (  9.38)    Acc@5  28.12 ( 28.12)
> 95
> Epoch: [96][0/3]    Time  0.724 ( 0.724)    Data  0.555 ( 0.555)    Loss 8.8189e+00 (8.8189e+00)    Acc@1   6.25 (  6.25)    Acc@5  15.62 ( 15.62)
> 96
> Epoch: [97][0/3]    Time  0.729 ( 0.729)    Data  0.559 ( 0.559)    Loss 8.8269e+00 (8.8269e+00)    Acc@1   6.25 (  6.25)    Acc@5  25.00 ( 25.00)
> 97
> Epoch: [98][0/3]    Time  0.781 ( 0.781)    Data  0.607 ( 0.607)    Loss 8.8340e+00 (8.8340e+00)    Acc@1   0.00 (  0.00)    Acc@5   0.00 (  0.00)
> 98
> Epoch: [99][0/3]    Time  0.747 ( 0.747)    Data  0.577 ( 0.577)    Loss 8.8407e+00 (8.8407e+00)    Acc@1   9.38 (  9.38)    Acc@5  34.38 ( 34.38)
> 99
> Training finished.

Then I tried to conduct the training using 12800 patches in the 25K patches (`epochs= 60, batch_size = 64`), the training process is still similar like above.

Could you provide some suggestions about how to solve this problem?

Thanks very much for your help.

**Comment by wangyi111** [comment]
Hi! There could be many reasons for the loss not converging. Apart from the hyper-parameter tuning, single-GPU training could also affect it. How did you modify shuffle batch normalization? MoCo officially supports only multi-GPU DistributedDataParallel to shuffle BN across GPUs. For a single GPU, it is possible to simulate that with SplitBatchNorm. This [colab demo on cifar10](https://colab.research.google.com/github/facebookresearch/moco/blob/colab-notebook/colab/moco_cifar10_demo.ipynb) might be helpful for your case.

**Comment by ftongss** [comment]
Thanks for your reply. 

I tried to simulate it with SplitBatchNorm, the training process is still same.

I made changes in two files `pretrain_moco_v2_s2c.py` and `builder.py`.

I upload the files I revised here [https://github.com/ftongss/SSL4EO-Test](url).

It is appreciated if you can have a look at the code.

The configuration to run the code is:
```
--data C:/Users/ftong/Desktop/lmdb/example_100_patches.lmdb
--checkpoints C:/Users/ftong/Desktop/lmdb/checkpoints_100_patches
--bands B13
--lmdb
--arch resnet50
--workers 0
--batch-size 32
--epochs 100
--lr 0.03
--mlp
--moco-t 0.2
--aug-plus
--cos
--seed 42
--gpu 0
--mode s2c
--dtype uint8
--season augment
--in_size 224
```
Thanks for your help.

**Comment by wangyi111** [comment]
Are you using only 100 patches? It's probably not very meaningful, as the training dataset size is too small. If you only want the loss to converge, maybe you can try:
- change the negative queue size, by default it is 65536 which exceeds 100 a lot.
```
parser.add_argument('--moco-k', default=65536, type=int, help='queue size; number of negative keys (default: 65536)')
```
- turn off the data augmentation
- start with a small learning rate
- etc..
